src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java

```diff
@@ -138,6 +140,8 @@ public Future<Path> startDownload(
                 eventHandler,
                 clientEnv,
                 context);
+          } finally {
+            doneSignal.countDown();
```

Comment:
this could lead to a deadlock if the future is never scheduled before being canceled (that is, after `submit()` is called and returns, but before the Callable gets called).

`WorkerSkyKeyComputeState` works around this by using a `ListeningExecutorService` and using `addListener` on the returned Future (https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/skyframe/WorkerSkyKeyComputeState.java;drc=23e5ec157d582585bb9635b5c374ec21cc50b4ea;l=173). let's do the same here

Reply:
That is a good point, but I verified, by both code inspection and empirically, that the `Future.addListener` also triggers its runnable too early. Therefore, I had to add a semaphore to get the ordering right, which my small scale experiment confirms work. (Well, I could not trigger the case where the task never starts, but the bug this is aimed to solve is gone.)